packages/lesswrong/components/ultraFeed/ultraFeedSettingsValidation.ts

```diff
@@ -1,56 +1,86 @@
 import { z } from 'zod';
 import { DEFAULT_SETTINGS } from './ultraFeedSettingsTypes';
+import { FeedItemSourceType, allFeedItemSourceTypes } from './ultraFeedTypes';
 
-const ascendingNumbersOrNull = z.array(z.number().positive().nullable()).refine(arr => {
+const sourceWeightsSchema = z.object(
+  Object.fromEntries(
+    allFeedItemSourceTypes.map(key => [
+      key,
+      z.number().min(0, { message: "Weight must be non-negative" })
+    ])
+  ) as Record<FeedItemSourceType, z.ZodNumber>
```

Comment:
avoid cast:
```
z.record(
  z.enum(allFeedItemSourceTypes),
  z.number().min(0, { message: "Weight must be non-negative" }),
)
```
?

Reply:
I can remove the cast. Otherwise, I've tried this with several variations and the existing thing is the only one that doesn't end up being partial and causing issues

Reply:
Well I can do this:

```
const sourceWeightsShape = allFeedItemSourceTypes.reduce((acc, key) => {
  acc[key] = z.number().min(0, { message: "Weight must be non-negative" });
  return acc;
}, {} as { [K in FeedItemSourceType]: z.ZodNumber });

const sourceWeightsSchema = z.object(sourceWeightsShape);```